File: object/object/src/Lec02/Screening.java
package Lec02;



import java.time.LocalDateTime;


/**
 * Screening 역할 : 사용자들이 예매하는 대상
 * Screening 책임 : 1. 예매
 *                 2. 환불
 *
 */
public class Screening {

    private Movie movie;
    private RefundCalculator refundCalculate;
    private int sequence;
    private LocalDateTime whenScreened;

    public Movie getMovie() {
        return movie;
    }

    public Screening(Movie movie, RefundCalculator refundCalculate, int sequence, LocalDateTime whenScreened) {
        this.movie = movie;
        this.refundCalculate = refundCalculate;
        this.sequence = sequence;
        this.whenScreened = whenScreened;
    }

    public LocalDateTime getStartTime(){
        return whenScreened;
    }

    public boolean isSequence(int sequence){
        return this.sequence == sequence;
    }

    public Money getMovieFee(){
        return movie.getFee();
    }

    public Reservation reserve(Customer customer, int audienceCount){
        return new Reservation(customer, this, calculateFee(audienceCount), audienceCount);
    }

    public Money calculateFee(int audienceCount){

        return movie.calculateMovieFee(this).times(audienceCount);
    }


    public Refund refund(Customer customer, int audienceCount){
        return new Refund(customer, this,calculateRefundFee(audienceCount),audienceCount);
    }

    private Money calculateRefundFee(int audienceCount){
        return refundCalculate.calculateRefundFee(this).times(audienceCount);

    }
}

File: object/object/src/java_to_kotlin/Screening.kt
package java_to_kotlin

import java.time.LocalDateTime

/**
 * Screening 역할 : 사용자들이 예매하는 대상
 * Screening 책임 : 1. 예매
 * 2. 환불
 *
 */
class Screening(val movie: Movie, private val sequence: Int, val startTime: LocalDateTime) {
    private val maxAudience: Int? = null
    fun isSequence(sequence: Int): Boolean {
        return this.sequence == sequence
    }

    val movieFee: Money?
        get() = movie.fee

    fun reserve(customer: Customer?, audienceCount: Int): Reservation {
        return Reservation(customer, this, calculateFee(audienceCount), audienceCount)
    }

    fun calculateFee(audienceCount: Int): Money? {
        return movie.calculateMovieFee(this)!!.times(audienceCount.toDouble())
    }

    fun refundCalculator(refund: Refund): Money? {
        return movie.calculateRefundFee(refund)
    }
}
